assets/shaders/viewport_render.wgsl

```diff
@@ -146,39 +145,44 @@ fn cube_intersect_ray(cube: Cube, ray: Line) -> CubeRayIntersection{
     }
 
     let p = point_in_ray_at_distance(ray, tmin);
-    var impact_normal = vec3f(0.,0.,0.);
-    if abs(p.x - cube.min_position.x) < FLOAT_ERROR_TOLERANCE {
-        impact_normal.x = -1.;
-    } else if abs(p.x - (cube.min_position.x + cube.size)) < FLOAT_ERROR_TOLERANCE {
-        impact_normal.x = 1.;
-    } else if abs(p.y - cube.min_position.y) < FLOAT_ERROR_TOLERANCE {
-        impact_normal.y = -1.;
-    } else if abs(p.y - (cube.min_position.y + cube.size)) < FLOAT_ERROR_TOLERANCE {
-        impact_normal.y = 1.;
-    } else if abs(p.z - cube.min_position.z) < FLOAT_ERROR_TOLERANCE {
-        impact_normal.z = -1.;
-    } else if abs(p.z - (cube.min_position.z + cube.size)) < FLOAT_ERROR_TOLERANCE {
-        impact_normal.z = 1.;
-    }
+
 
     if tmin < 0.0 {
         result.hit = true;
         result.impact_hit = false;
         result.exit_distance = tmax;
-        result.impact_normal = impact_normal;
         return result;
     }
 
     result.hit = true;
     result.impact_hit = true;
     result.impact_distance = tmin;
     result.exit_distance = tmax;
-    result.impact_normal = impact_normal;
     return result;
 }
 
+fn cube_impact_normal(cube: Cube, impact_point: vec3f) -> vec3f{
+    var impact_normal = vec3f(0.,0.,0.);
+    let mid_to_impact = cube.min_position + vec3f(cube.size / 2.) - impact_point;
+    let mid_to_impact_abs = abs(mid_to_impact);
+    let max_component = max(
+        mid_to_impact_abs.x,
+        max(mid_to_impact_abs.y, mid_to_impact_abs.z)
+    );
+    if max_component - mid_to_impact_abs.x < FLOAT_ERROR_TOLERANCE {
+        impact_normal.x = -mid_to_impact.x;
+    }
+    if max_component - mid_to_impact_abs.y < FLOAT_ERROR_TOLERANCE {
+        impact_normal.y = -mid_to_impact.y;
+    }
+    if max_component - mid_to_impact_abs.z < FLOAT_ERROR_TOLERANCE {
+        impact_normal.z = -mid_to_impact.z;
+    }
+    return normalize(impact_normal);
+}
```

Comment:
Additionaly to moving this into its own function, the logic changed as well to make it more robust; 

The normal vector is calculated from its maximum component.